HW0/P5/P5.txt

```diff
@@ -1 +1,12 @@
 # Your answers here.
+1, 255 seconds
+2, 34 seconds. 8 people and each people takes 32 bags. It takes 31 seconds to finish the counting at the first stage. Furthremore, it takes log2(8) = 3 seconds to finish. So in total it is 31+3 = 34 seconds.
+3, 8 seconds
+4, done. Its log2(N)(round up to the nearest integer) seconds
+5, computing time = 8 seconds. However the communication time is 7 seconds, because at the first stage, the cashier would not need to communicate with himself. so in total it is 15 seconds.
+6, If each cashier comes and takes 1 bag. It takes 256 + 8 + 8 = 272 seconds.
+if each cashier comes and takes 2 bags. It takes 128 + 8 +7 to finish
+and yes, there is better strategies to shorten the time.
+
+For me the best way to is pick 23 to the first guy, and 22 to the second guy, etc.. 
```

Comment:
I think the fastest way takes 22 seconds.
1st second: 256 bags to cashier 1
2nd second: cashier 1 gives 128 bags to cashier 2
3rd second: cashier 1 and cashier 2 give 64 bags to cashier 3 and cashier 4
4th second: cashier 1, 2, 3, 4 give 32 bags to cashier 5, 6, 7, 8
5th second: cashier 1...8 give 16 bags to cashier 7...16
6th second: cashier 1...16 give 8 bags to cashier 17...32
7th second: cashier 1...32 give 4 bags to cashier 33...64
10th second: every cashier counts 4 bags in 3 seconds.
Now we have 32 pairs of values.
11th second: communicate among 64 cashiers
12th second: adds up and ends up with 32 values
13th second: communicate again
14th second: adds up and ends up with 16 values
15th second: communicate again
16th second: adds up and ends up with 8 values
17th second: communicate again
18th second: adds up and ends up with 4 values
19th second: communicate again
20th second: adds up and ends up with 2 values
21st second: communicate again
22nd second: 1 value. Done!